### ISSUE TITLE
feat - color mode and advanced analysis

### ISSUE THREAD

**Issue by tintinweb** [tintinweb/ethereum-dasm#3]
## advanced analysis

* gas calculation
* arguments and return values
* function name guessing
* payable modifier detection
* argument number and length detection

## color mode

![image](https://user-images.githubusercontent.com/2865694/42599610-11334024-8560-11e8-92d4-f09bf2e67ae1.png)

## non-color mode

--> scroll right

```
Inst   addr  hex   gas | mnemonic        operand                              xrefs                          description
------------------------------------------------------------------------------------------------------------------------------------------------------
:loc_0x0
   0 [   0 0x0000]   3 | PUSH1           0x60                                                                # Place 1 byte item on stack.                                  returns: item
   1 [   2 0x0002]   3 | PUSH1           0x40                                                                # Place 1 byte item on stack.                                  returns: item
   2 [   4 0x0004]   3 | MSTORE                                                                              # Save word to memory.                                                                        args: value, offset
   3 [   5 0x0005]   3 | PUSH1           0x04                                                                # Place 1 byte item on stack.                                  returns: item
   4 [   7 0x0007]   2 | CALLDATASIZE                                                                        # Get size of input data in current environment.               returns: msg.data.length
   5 [   8 0x0008]   3 | LT                                                                                  # Lesser-than comparison                                       returns: flag                  args: a, b
   6 [   9 0x0009]   3 | PUSH2           0x0048                                                              # Place 2-byte item on stack.                                  returns: item
   7 [  12 0x000c]  10 | JUMPI           @0x48                                                               # Conditionally alter the program counter.                                                    args: evm.pc, condition
   8 [  13 0x000d]   3 | PUSH4           0xffffffff                                                          # Place 4-byte item on stack.                                  returns: item
   9 [  18 0x0012]   3 | PUSH1           0xe0                                                                # Place 1 byte item on stack.                                  returns: item
  10 [  20 0x0014]   3 | PUSH1           0x02                                                                # Place 1 byte item on stack.                                  returns: item
  11 [  22 0x0016]  10 | EXP                                                                                 # Exponential operation.                                       returns: result                args: base, exponent
  12 [  23 0x0017]   3 | PUSH1           0x00                                                                # Place 1 byte item on stack.                                  returns: item
  13 [  25 0x0019]   3 | CALLDATALOAD                                                                        # Get input data of current environment.                       returns: msg.data              args: unknown
  14 [  26 0x001a]   5 | DIV                                                                                 # Integer division operation.                                  returns: result                args: a, b
  15 [  27 0x001b]   3 | AND                                                                                 # Bitwise AND operation.                                       returns: result                args: a, b
  16 [  28 0x001c]   3 | PUSH4           0x124c32a1  --> 'function enter(bytes32,bytes8)'                                # Place 4-byte item on stack.                                  returns: item
  17 [  33 0x0021]   3 | DUP2                                                                                # Duplicate 2nd stack item.
  18 [  34 0x0022]   3 | EQ                                                                                  # Equality  comparison                                         returns: flag                  args: a, b
  19 [  35 0x0023]   3 | PUSH2           0x004d                                                              # Place 2-byte item on stack.                                  returns: item
  20 [  38 0x0026]  10 | JUMPI           @0x4d                                                               # Conditionally alter the program counter.                                                    args: evm.pc, condition
  21 [  39 0x0027]   3 | DUP1                                                                                # Duplicate 1st stack item.
  22 [  40 0x0028]   3 | PUSH4           0x60643652  --> 'function maxEntrants()'                                # Place 4-byte item on stack.                                  returns: item
  23 [  45 0x002d]   3 | EQ                                                                                  # Equality  comparison                                         returns: flag                  args: a, b
  24 [  46 0x002e]   3 | PUSH2           0x0095                                                              # Place 2-byte item on stack.                                  returns: item
  25 [  49 0x0031]  10 | JUMPI           @0x95                                                               # Conditionally alter the program counter.                                                    args: evm.pc, condition
  26 [  50 0x0032]   3 | DUP1                                                                                # Duplicate 1st stack item.
  27 [  51 0x0033]   3 | PUSH4           0x694463a2  --> 'function totalEntrants()'                                # Place 4-byte item on stack.                                  returns: item
  28 [  56 0x0038]   3 | EQ                                                                                  # Equality  comparison                                         returns: flag                  args: a, b
  29 [  57 0x0039]   3 | PUSH2           0x00be                                                              # Place 2-byte item on stack.                                  returns: item
  30 [  60 0x003c]  10 | JUMPI           @0xbe                                                               # Conditionally alter the program counter.                                                    args: evm.pc, condition
  31 [  61 0x003d]   3 | DUP1                                                                                # Duplicate 1st stack item.
  32 [  62 0x003e]   3 | PUSH4           0x90ae631d  --> 'function assignAll()'                                # Place 4-byte item on stack.                                  returns: item
  33 [  67 0x0043]   3 | EQ                                                                                  # Equality  comparison                                         returns: flag                  args: a, b
  34 [  68 0x0044]   3 | PUSH2           0x00d1                                                              # Place 2-byte item on stack.                                  returns: item
  35 [  71 0x0047]  10 | JUMPI           @0xd1                                                               # Conditionally alter the program counter.                                                    args: evm.pc, condition

:loc_0x48
  36 [  72 0x0048]   1 | JUMPDEST                                             JUMPI@0xc                      # Mark a valid destination for jumps.
  37 [  73 0x0049]   3 | PUSH1           0x00                                                                # Place 1 byte item on stack.                                  returns: item
  38 [  75 0x004b]   3 | DUP1                                                                                # Duplicate 1st stack item.
  39 [  76 0x004c]   0 | REVERT                                                                              # throw an error                                                                              args: offset, size
:loc_0x4d
  /*******************************************************************
    function enter(bytes32,bytes8)
    payable: False
    inputs: (2) ['bytes32', '<bytes??>']
    potential signatures: ['enter(bytes32,bytes8)']
  *******************************************************************/

  40 [  77 0x004d]   1 | JUMPDEST                                             JUMPI@0x26                     # Mark a valid destination for jumps.
  41 [  78 0x004e]   2 | CALLVALUE                                                                           # Get deposited value by the instruction/transaction responsible for this execution. returns: msg.value
  42 [  79 0x004f]   3 | ISZERO                                                                              # Simple not operator                                          returns: flag                  args: a
  43 [  80 0x0050]   3 | PUSH2           0x0058                                                              # Place 2-byte item on stack.                                  returns: item
  44 [  83 0x0053]  10 | JUMPI           @0x58                                                               # Conditionally alter the program counter.                                                    args: evm.pc, condition
  45 [  84 0x0054]   3 | PUSH1           0x00                                                                # Place 1 byte item on stack.                                  returns: item
  46 [  86 0x0056]   3 | DUP1                                                                                # Duplicate 1st stack item.
  47 [  87 0x0057]   0 | REVERT                                                                              # throw an error                                                                              args: offset, size
:loc_0x58
  48 [  88 0x0058]   1 | JUMPDEST                                             JUMPI@0x53                     # Mark a valid destination for jumps.
  49 [  89 0x0059]   3 | PUSH2           0x0081                                                              # Place 2-byte item on stack.                                  returns: item
  50 [  92 0x005c]   3 | PUSH1           0x04                                                                # Place 1 byte item on stack.                                  returns: item
  51 [  94 0x005e]   3 | CALLDATALOAD                                                                        # Get input data of current environment.                       returns: msg.data              args: unknown
  52 [  95 0x005f]   3 | PUSH24          0xffffffffffffffffffffffffffffffffffffffffffffffff                                # Place 24-byte item on stack.                                 returns: item
  53 [ 120 0x0078]   3 | NOT                                                                                 # Bitwise NOT operation.                                       returns: result                args: a, b
  54 [ 121 0x0079]   3 | PUSH1           0x24                                                                # Place 1 byte item on stack.                                  returns: item
  55 [ 123 0x007b]   3 | CALLDATALOAD                                                                        # Get input data of current environment.                       returns: msg.data              args: unknown
  56 [ 124 0x007c]   3 | AND                                                                                 # Bitwise AND operation.                                       returns: result                args: a, b
  57 [ 125 0x007d]   3 | PUSH2           0x00e4                                                              # Place 2-byte item on stack.                                  returns: item
  58 [ 128 0x0080]   8 | JUMP            @0xe4                                                               # Alter the program counter.                                                                  args: evm.pc

:loc_0x81
  59 [ 129 0x0081]   1 | JUMPDEST                                                                            # Mark a valid destination for jumps.
  60 [ 130 0x0082]   3 | PUSH1           0x40                                                                # Place 1 byte item on stack.                                  returns: item
  61 [ 132 0x0084]   3 | MLOAD                                                                               # Load word from memory.                                                                      args: offset
  62 [ 133 0x0085]   3 | SWAP1                                                                               # Exchange 1st and 2nd stack items.
  63 [ 134 0x0086]   3 | ISZERO                                                                              # Simple not operator                                          returns: flag                  args: a
  64 [ 135 0x0087]   3 | ISZERO                                                                              # Simple not operator                                          returns: flag                  args: a
  65 [ 136 0x0088]   3 | DUP2                                                                                # Duplicate 2nd stack item.
  66 [ 137 0x0089]   3 | MSTORE                                                                              # Save word to memory.                                                                        args: value, offset
  67 [ 138 0x008a]   3 | PUSH1           0x20                                                                # Place 1 byte item on stack.                                  returns: item
  68 [ 140 0x008c]   3 | ADD                                                                                 # Addition operation.                                          returns: result                args: a, b
  69 [ 141 0x008d]   3 | PUSH1           0x40                                                                # Place 1 byte item on stack.                                  returns: item
  70 [ 143 0x008f]   3 | MLOAD                                                                               # Load word from memory.                                                                      args: offset
  71 [ 144 0x0090]   3 | DUP1                                                                                # Duplicate 1st stack item.
  72 [ 145 0x0091]   3 | SWAP2                                                                               # Exchange 1st and 3rd stack items.
  73 [ 146 0x0092]   3 | SUB                                                                                 # Subtraction operation.                                       returns: result                args: a, b
  74 [ 147 0x0093]   3 | SWAP1                                                                               # Exchange 1st and 2nd stack items.
  75 [ 148 0x0094]   0 | RETURN                                                                              # Halt execution returning output data.                                                       args: offset, size

:loc_0x95
  /*******************************************************************
    function maxEntrants()
    payable: False
    inputs: (0) []
    potential signatures: ['maxEntrants()']
  *******************************************************************/
...
```

## color mode

* the alignment is off

[![asciicast](https://asciinema.org/a/CXLtlreNQ4G52KJqDfKcDGg1X.png)](https://asciinema.org/a/CXLtlreNQ4G52KJqDfKcDGg1X)

**Comment by tintinweb** [comment]
* new console tablelib
* symbolic execution

![screenshot windows](https://user-images.githubusercontent.com/2865694/42905573-c0b9f2cc-8ad8-11e8-9349-651e1116b6cc.png)

**Comment by tintinweb** [comment]
closed by #4
